calyx/src/passes/inliner.rs

```diff
@@ -22,139 +23,124 @@ impl Named for Inliner {
-    for idx in st
-        .edge_idx()
-        .with_direction(DataDirection::Write)
-        .with_node_type(NodeType::Hole)
-        .with_port(hole)
-        .detach()
-    {
-        st.remove_edge(idx);
-    }
-
-    // flatten groups (maybe temporary)
-    for idx in st.edge_idx().detach() {
-        st.get_edge_mut(idx).group = None;
-    }
-    st.groups = HashMap::new();
-    st.groups
-        .insert(None, (HashMap::new(), st.edge_idx().collect()));
-}
+        // add top_level[go] = this.go
+        let go_asgn = Assignment {
+            src: builder.add_constant(1, 1).borrow().get("out"),
+            dst: top_level.borrow().get("go"),
+            guard: Some(Guard::Port(go_sig)),
+        };
+        // add this.done = top_level[done]
+        let done_asgn = Assignment {
+            src: builder.add_constant(1, 1).borrow().get("out"),
+            dst: done_rig,
+            guard: Some(Guard::Port(top_level.borrow().get("done"))),
+        };
+        comp.continuous_assignments.push(go_asgn);
+        comp.continuous_assignments.push(done_asgn);
+
+        // construct analysis graph and find sub-graph of all edges that include a hole
+        let analysis = Analysis::from(&comp);
+        let subgraph = analysis
+            .clone()
+            .edge_induced_subgraph(|src, dst| src.is_hole() || dst.is_hole());
+
+        // if subgraph has cycles, error out
+        if subgraph.has_cycles() {
+            // XXX use topo sort to find where the cycle is
+            return Err(Error::MalformedStructure(
+                "Cyclic hole definition.".to_string(),
+            ));
+        }
 
-impl Visitor for Inliner {
-    fn start(&mut self, comp: &mut Component, _c: &Context) -> VisResult {
-        let st = &mut comp.structure;
+        // map of holes to their guard expressions
+        let mut map = HashMap::new();
+        let mut assignments = vec![];
+        for group in &comp.groups {
+            // compute write closure
+            for hole in &group.borrow().holes {
+                map.insert(
+                    hole.borrow().key(),
+                    write_closure(&analysis, &hole.borrow()),
+                );
+            }
+
+            // remove all assignments from group, taking ownership
+            let mut group = group.borrow_mut();
+            assignments.append(&mut group.assignments.drain(..).collect());
+        }
 
-        // inline `go` holes first because `done` holes may reference them
-        inline_hole(st, "go".to_string());
-        inline_hole(st, "done".to_string());
+        // add the continuous assignment edges
+        assignments
+            .append(&mut comp.continuous_assignments.drain(..).collect());
+
+        // remove edges that write to a hole
+        assignments.retain(|asgn| !asgn.dst.borrow().is_hole());
+
+        // move direct reads from holes into the guard so they can be inlined
+        let mut builder = Builder::from(comp, sigs, false);
```

Comment:
Isn't there a builder already in scope?

Reply:
builder requires mutable access to comp as long as it is alive which prevents other mutable borrows. Reconstructing it shortens the lifespan of the builder. I'll add some comments to make this more clear

Reply:
`builder.component` returns a mutable reference to the component. I've used this pattern once in `compile_control`. Maybe we can have a function that operates on the mutable component reference defined on the `builder`?